pkg/agent/openflow/pipeline.go

```diff
@@ -598,37 +623,135 @@ func (c *client) connectionTrackFlows(category cookie.Category) []binding.Flow {
+			Done())
+
+		for _, proto := range c.ipProtocols {
+			gatewayIP := c.nodeConfig.GatewayConfig.IPv4
+			serviceGWHairpinIP := config.ServiceGWHairpinIPv4.To4()
+			serviceCtZone := ServiceCtZone
+			ctZone := CtZone
+			if proto == binding.ProtocolIPv6 {
+				gatewayIP = c.nodeConfig.GatewayConfig.IPv6
+				serviceGWHairpinIP = config.ServiceGWHairpinIPv6.To16()
+				serviceCtZone = ServiceCtZoneV6
+				ctZone = CtZoneV6
+			}
+			flows = append(flows,
+				// This flow is used to maintain DNAT conntrack for Service traffic.
+				connectionTrackTable.BuildFlow(priorityNormal).MatchProtocol(proto).
+					Action().CT(false, connectionTrackTable.GetNext(), ctZone).NAT().CTDone().
+					Cookie(c.cookieAllocator.Request(category).Raw()).
+					Done(),
+				connectionTrackCommitTable.BuildFlow(priorityLow).MatchProtocol(proto).
+					MatchCTStateTrk(true).
+					MatchCTMark(ServiceCTMark, nil).
+					MatchRegRange(int(serviceLearnReg), marksRegServiceSelected, serviceLearnRegRange).
+					Cookie(c.cookieAllocator.Request(category).Raw()).
+					Action().GotoTable(connectionTrackCommitTable.GetNext()).
+					Done(),
+			)
+
+			if c.enableProxyFull {
+				serviceConnectionTrackTable := c.pipeline[serviceConntrackTable]
+				serviceConnectionTrackCommitTable := c.pipeline[serviceConntrackCommitTable]
+				flows = append(flows,
+					// This flow is used to match the Service traffic from Antrea gateway. The Service traffic from gateway
+					// should enter table serviceConntrackCommitTable, otherwise it will be matched by other flows in
+					// table connectionTrackCommit.
+					connectionTrackCommitTable.BuildFlow(priorityHigh).MatchProtocol(proto).
+						MatchCTMark(ServiceCTMark, nil).
+						MatchRegRange(int(marksReg), markTrafficFromGateway, binding.Range{0, 15}).
+						Action().GotoTable(serviceConntrackCommitTable).
+						Cookie(c.cookieAllocator.Request(category).Raw()).
+						Done(),
+					// This flow is used to maintain SNAT conntrack for Service traffic.
+					serviceConnectionTrackTable.BuildFlow(priorityNormal).MatchProtocol(proto).
+						Action().CT(false, serviceConnectionTrackTable.GetNext(), serviceCtZone).NAT().CTDone().
+						Cookie(c.cookieAllocator.Request(category).Raw()).
+						Done(),
+					// This flow is used to match the first packet of NodePort/LoadBalancer whose Endpoint is on
+					// host network. As the packet is from Antrea gateway, and it will pass through Antrea gateway,
+					// a virtual hairpin IP is used to perform SNAT for the packet, rather than Antrea gateway's IP.
+					// Note that, this flow will change the behavior of the packet that NodePort/LoadBalancer whose
+					// externalTrafficPolicy is Local and the Endpoint is on host network. According to the definition
+					// of externalTrafficPolicy Local, the source IP should be retained.  If a pod is on host network,
+					// a Node cannot hold more than one pod like the pod. There is no point to expose the pod as NodePort,
```

Comment:
"more than one Pods of the Service"?

Reply:
This is the new comment

```
If the Endpoint is on host network,
// there will be only one Pod (It is impossible to have more than one Pods which listen on the same
// port on host network), so there is no point to expose the Pod as NodePort, as it makes no difference
// to access it directly.
```

Reply:
Suggestions:
* "there will be" -> "there should be"
* "only one Pod" -> "only one backend Pod of the Service on a Node".
* "there is no point" -> "it is not useful". "no point" is too strong in my mind.

Question - in theory two Pods can define the same named port with two different port number, and the NodePort Service can use the named port to be the target port? I do not mean it is common and we should support. Just asking.

Reply:
Reminder.

Reply:
Thanks.